lib/timers.js

```diff
@@ -87,13 +87,20 @@ function listOnTimeout() {
 
   var first;
   while (first = L.peek(list)) {
-    // If the previous iteration caused a timer to be added,
-    // update the value of "now" so that timing computations are
-    // done correctly. See test/simple/test-timers-blocking-callback.js
-    // for more information.
+    // Whenever a timer creates a new timer in the current list being processed
+    // you can end up processing the new timer immediately instead of waiting
+    // the scheduled time.  Whenever we encounter this situation stop
+    // processing this list and reschedule.
+    //
+    // See: https://github.com/joyent/node/issues/25607
     if (now < first._monotonicStartTime) {
-      now = Timer.now();
-      debug('now: %d', now);
+      var delay = msecs - (Timer.now() - first._monotonicStartTime);
```

Comment:
I would suggest renaming `delay` to `timeRemaining`, as it seems closer to what this value represents, but I'm not a native English speaker, so I may be wrong.